implement preloading modules

lucebac opened the pull request.

gaelmuller commented:
Thank you for this PR. It is easier to review 😄 

I will review it in the next few days.

gaelmuller commented on docs/modules.rst:
```diff
@@ -28,7 +29,53 @@ The best practice is to do the following:
+Adding preloading results
+-------------------------
+
+Once the module successfully preloaded the sample for FAME, it must add the result to the analysis. As soon as such result is added, FAME schedules the regular analysis based on what was added to the analysis.
```
I find this to be confusing. I would remove the paragraph.

lucebac replied:
Anything specific that makes you feel confused? Maybe this just needs to rephrased.

gaelmuller replied:
I guess it is because is mentions "results" in a generic way when preloading modules can really have only one output: a file.

lucebac replied:
I rephrased the section, could you please re-read?

gaelmuller commented on web/templates/modules/index.html:
```diff
@@ -100,6 +100,72 @@ <h3 class="panel-title">Error</h3>
+                            <ul>
+                                {% if module.acts_on %}
+                                    <li><span class="label">Acts On</span> {{module.acts_on|smart_join}}</li>
+                                {% endif %}
```
I don't think Preloading Modules have an acts_on

lucebac replied:
Slipped through while simplifying the preloading module implementation. Will remove it.

gaelmuller commented on web/templates/modules/module_configuration.html:
```diff
@@ -19,7 +19,24 @@ <h4 class="title">{{data.module.name}}</h4>
+                                <input type="text" class="form-control" name="acts_on" value="{{data.module.acts_on|smart_join}}">
+                                <p class="help-block">Comma-delimited list of types that this module is able to handle (ex: word,excel).</p>
+                            </div>
+                        </div>
```
I don't think Preloading Modules have an acts_on

lucebac replied:
Slipped through while simplifying the preloading module implementation. Will remove it.

gaelmuller commented on web/views/analyses.py:
```diff
@@ -351,7 +324,7 @@ def get_file(self, id, filehash):
                 filepath = filepath.encode('utf-8')
                 if filehash == md5(filepath).hexdigest():
-                    return file_download(filehash)
+                    return file_download(filepath)
```
Not sure what this change is doing here. I don't think this is related to preloading modules so it should be removed.

lucebac replied:
No, this was an unrelated bugfix. Will remove it from this PR.

lucebac commented on fame/core/module_dispatcher.py:
```diff
@@ -47,12 +48,26 @@ def reload(self):
+        return (
+            self.get_preloading_module(module_name) or
+            self.get_processing_module(module_name) or
+            None
```
In general, you are right. But I, personally, would like to have the `return None` explicitly written down here.

gaelmuller replied:
Fair enough :+1:

lucebac commented on fame/core/module.py:
```diff
@@ -1108,3 +1113,68 @@ def static_info(cls):
+        self.init_options(analysis['options'])
+        return self.preload(self._analysis.get_main_file())
+
+    def log(self, level, message):
```
sure, it's always the same code anyway

lucebac commented on fame/core/module.py:
```diff
@@ -1108,3 +1113,68 @@ def static_info(cls):
+            target (string): the hash that is to be analyzed
+        Raises:
+            ModuleExecutionError: Preloading the analysis failed (e.g.
+                                  no file for a given hash was found).
```
This is an inconsistency. Somewhere else in the docu I stated that the module should return `False` if no file could be downloaded (and `True` if download was successful). Will remove the exception.

lucebac commented on docs/concept.rst:
```diff
@@ -79,8 +80,8 @@ FAME relies on three components:
 .. image:: /images/concept-architecture.png
 
-Components can all be on the same server, or split across multiple servers.
+Components can all be on the same server, or split across multiple servers. A dockerized version of FAME is also available.
```
That sentence slipped through while copy-pasting the changes. Will remove it.

lucebac commented on docs/modules.rst:
```diff
@@ -28,7 +29,53 @@ The best practice is to do the following:
-* Inside your python file, create a class that inherits from :class:`fame.core.module.ProcessingModule`, :class:`fame.core.module.ReportingModule`, :class:`fame.core.module.ThreatIntelligenceModule` or :class:`fame.core.module.AntivirusModule`.
+* Inside your python file, create a class that inherits from :class:`fame.core.module.PreloadingModule`, :class:`fame.core.module.ProcessingModule`, :class:`fame.core.module.ReportingModule`, :class:`fame.core.module.ThreatIntelligenceModule` or :class:`fame.core.module.AntivirusModule`.
+
+Writing a Preloading module
```
Could you explain that a little bit? In my opinion (and compared to in what order the modules will be executed) it makes more sense to put Preloading before Processing simply because the former is executed before the latter.

gaelmuller replied:
Based on execution order, you are correct, but I think that most people looking to develop modules are looking for Processing Modules. 

In my opinion, it is the main kind of modules so it should always come first.

lucebac replied:
Agreed

lucebac commented:
Thanks for the review. I added one or another question to some of your comments for which I need clarification/more details.

gaelmuller commented on docs/modules.rst:
```diff
@@ -28,7 +29,53 @@ The best practice is to do the following:
+Testing Preloading Modules
+--------------------------
+
+See :ref:`testing_processing_modules`.
```
I think this should be removed since there is currently no way to test a Preloading Module with `single_module.py`.

lucebac replied:
removed

gaelmuller commented on fame/core/analysis.py:
```diff
@@ -182,6 +184,40 @@ def add_ioc(self, value, source, tags=[]):
+        self._file = f
+        self.save()
+
+        self._automatic(preloading_done=True)
```
If I understand correctly, this is creating a new File object with the actual file. I would expect the "hash" file to be replaced instead.

lucebac replied:
Under the hood, this is exactly what we tried to implement. But as far as I remember, we could not (easily) update the currently used `File` object and opted for just creating a new one and replacing it in the `Analysis` object. I can check in the code if I now see a proper way to implement that - but I would like to postpone that until we are done with all the (upcoming) PRs.

gaelmuller commented on fame/core/file.py:
```diff
@@ -11,14 +11,58 @@
+            else:
+                self._compute_default_properties(hash_only=True)
+                self._init_hash(hash)
+                self.save()
```
The method is starting to get big. Maybe this could all go in a submethod ?

lucebac replied:
Yup, this is definitely getting big. I'll introduce a submethod for the hash initialization code.

gaelmuller commented on fame/core/module.py:
```diff
@@ -1108,3 +1113,59 @@ def static_info(cls):
+    def execute(self, analysis):
+        self._analysis = analysis
+        self.init_options(analysis['options'])
+        return self.preload(self._analysis.get_main_file())
```
There should be a try catch here so that any exception happening in a preloading module does not prevent the other preloading modules to execute.

lucebac replied:
I added some try/catch logic

gaelmuller commented on web/templates/modules/index.html:
```diff
@@ -100,6 +100,68 @@ <h3 class="panel-title">Error</h3>
+        </div>
+    </div>
+
+
```
I think Preloading modules should be after Processing Modules

lucebac replied:
moved

gaelmuller commented on utils/single_module.py:
```diff
@@ -215,7 +215,7 @@ def test_mode_module(name, interactive):
         try:
             fame_init()
-            module = dispatcher.get_processing_module(args.module)
+            module = dispatcher.get_module(args.module)
```
All changes inside this file should be reverted since this PR is not providing support for preloading modules in `single_module.py`

lucebac replied:
file reverted

lucebac commented on fame/core/analysis.py:
```diff
@@ -293,7 +340,10 @@ def filepath(self, path):
+        filepath = self._file['filepath']
+        if self._file['type'] == "hash":
+            return filepath
+        return self.filepath(filehash)
```
fixed

lucebac commented:
>     * Preloading modules do not work when the analyst is specifying modules to execute. Preloading modules also appear in the modules list which is probably not what we want.
Regarding the first sentence: if I understand correctly what you are pointing out here, we also observed this behavior in the past (differently, but still). But this seems to be a more general problem with how modules are scheduled in FAME: a module will only be executed if FAME has a direct execution path to the module at the time it needs to decide which modules to execute (cf. `ModuleDispatcher:next_module`). To be honest: I do not want to touch this code since the algorithm is rather complex and I barely at all understand what is going on there.
In regards to the second sentence: I don't agree with your suggestion that preloading modules should not show up in the modules list. In my opinion the analyst should have the possibilty to also schedule preloading modules (e.g. if one wants to exclude/force specific preloading modules). This would not be possible if we did not show them in the module list.
 
>     * When preloading fails, it would be nice to have log messages explaining what happened.
Isn't this part of the reponsibilty of the module itself? I don't see a reason why FAME should guess why a certain module did not succeed when the module could instead just issue a log message stating what went wrong (like I did for the `virustotal_download` module). Nevertheless, I added a hint regarding that to the documentation for now.

>     * It would be nice to be able to define the order in which the preloading modules should be executed with a priority setting or something similar (optional, but nice to have so that we always prefer internal databases to costly external services).
Agreed. You already suggested that in the big PR and I put that on my todo list. But I'd like to postpone new features until we cleaned up the PR mess that I created.

lucebac commented:
Any news on this?

gaelmuller commented:
Sorry about the delay. I am still having the same issues with this feature at this point. I will make some changes, could you please make sure I have the rights to push commits on this branch ?

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

lucebac commented:
I don't see the checkbox described by the article you linked. I'm not sure whether there is another way to give you the ability to make changes to the PR.

gaelmuller commented:
I've created a new PR with my changes: https://github.com/certsocietegenerale/fame/pull/86

Could you please test / review it ?

lucebac commented:
closed in favor of #86
